### DISCUSSION TITLE
Discuss if / how to put computational outputs into content blocks (notes, warnings, etc)

### DISCUSSION THREAD

**Discussion by choldgraf** [executablebooks/meta#159] [General]
This has been discussed elsewhere (https://github.com/executablebooks/jupyter-book/issues/700, https://github.com/executablebooks/sphinx-exercise/issues/18 for example) so moving here to have a single place to discuss this topic.

# Current situation

Currently, executable code must exist in the form of a `{code-cell}` at the top level of a markdown file, or in the cell of a Jupyter Notebook. This is because we want a 1-to-1 mapping of text files <--> ipynb file via Jupytext, and since Jupyter Notebooks have a flat, top-level list of code cells, that is the only pattern that we support.

# The problem

The problem that others have noted is that you often want to embed the results of running code inside other parts of your content. The most common example that has come up is to put code cells / outputs inside of an admonition. For example:

````
:::{note}
Some note text.

```{code-cell}
print("this is a code cell")
```
:::
````

The above pattern is not possible currently, because `{code-cell}` needs to be a top-level item, mimicking the structure of a Jupyter Notebook (which has no concept of code cells nested in other cells).

So one question we'd need to answer to support this is: how would the above code block be translated into an `ipynb` file? Right now I do not believe that it is possible with any existing tool.

# A few potential solutions

Here are some ideas I can think of that we could discuss

1. **Make it possible to nest code cells via cell metadata**. In our Jupytext converter, we could support cell-level metadata like `admonition: note`, `admonition-body: Some body text`, `admonition-footer`, and when converted to markdown it would use this metadata to wrap the resulting `code-cell` in a `{note}` admonition.
   - The challenge with this is that it is very edge-casey...we're basically special-casing some cell metadata that would break down as soon as people wanted anything more complex, and it's also clunky to ask people to do complex configuration via cell metadata
2. **Make it easier to inject code cells and their outputs into a page**. Right now we have `glue`, a python function that tags a variable with a name for insertion. Others have pointed out this feels too cumberson, and is also python-specific. Perhaps making it easier to insert code cells / outputs from other pages will make this a workable solution.
   - The challenge with this is that we'd still require a two-step process for people to insert outputs etc. So maybe would still be too-cumbersome
   - See https://github.com/executablebooks/MyST-NB/issues/265 for some discussion.
3. **Allow for in-page execution that **doesn't** map onto notebook cells**. For example, we could introduce MyST-NB-specific syntax that would execute in the context of a Sphinx build, but would not directly translate to an `ipynb` file (it would just be copied as raw text or something).
   - For example, something like `` {execute}`print("hi")` `` would be run when the book was built, but would be meaningless in the context of a Jupyter interface.
   - The challenge with this is that we are currently using `nbclient` for our execution, and this doesn't support anything like what I describe.
   - We could consider using something like `papermill` and writing a post-execution hook that would run these blocks of code **after** executing all of the cells on the page.
   - This might be possible in `nbclient` if we had something like https://github.com/jupyter/nbclient/issues/81
- **Provide a way for extensions to hack this behavior** - I guess the last thing we could do is to figure out how an extension like `sphinx-exercise` could work around this behavior (maybe by adopting its own special cell metadata?) and then document how other extensions could do the same thing.
   - The challenge with this is that it's also very edge-casey and may not be possible anyway...

I don't think any of these solutions feels obviously correct to me. Personally if the machinery were there to make 3 possible, I think that's the one I'd prefer since it is closest to what RMarkdown does, and people seem to quite like that feature.

Maybe others have ideas for supporting this through other means - I would love to hear your thoughts.

**Comment by chrisjsewell** [comment]
> Allow for in-page execution that doesn't map onto notebook cells. 

My thought is this is an all or nothing option: the files either map 1:1 markdown <-> notebook, or they are not associated with notebooks at all.
You can't have parts of the code that don't work in the notebook context, that would be a complete mess.
If you want to go with this option, we should just use jupyter-sphinx or something similar. It's a valid option, but note we will have to drop notebook specific functionality like Binder and Colab, plus a new/adapted caching system (as opposed to jupyter-cache), plus plus users would no longer be able to use pre-executed notebooks

(2) is the better option IMO

**Reply by choldgraf** [reply]
Yeah, I agree with the main drawback of option 3 - we're basically introducing two kinds of computational behavior depending on whether a file is run in a myst-nb context or a jupyter book context. Also agreed that we're basically talking about jupyter-shpinx (another way I'd describe option 2 is essentially "allow jupyter sphinx and myst-nb to share a kernel).

That said I still wonder if it is the lesser-of-all-evils, partially because this is already a mental map that people coming from the R community will have and that has proven to be very popular.

However, if we can find a pattern that follows 2 and that is both simple in UX and implementation, I agree that this would be the preferable option.

It would be helpful to hear from @jstac and the QuantEcon team as this seems to be a particularly important topic for them.

**Comment by akhmerov** [comment]
Also linking https://github.com/executablebooks/meta/issues/143 with the same question being discussed. I think the main relevant point from that discussion is that it would be possible to support code cells nested anywhere, if jupytext conversion was based on myst-parser as well as sphinxcontrib-rst2myst and not only markdown-it-py, but this has unreasonably high implementation costs.

Out of the options above, I think (2) (a.k.a. https://github.com/executablebooks/MyST-NB/issues/265) is indeed the most reasonable since it's completely general, and doesn't break user assumptions about execution order etc.

**Reply by chrisjsewell** [reply]
> it would be possible to support code cells nested anywhere, if jupytext conversion was based on sphinxcontrib-rst2myst and not markdown-it-py

As I've pointed out in that issue, this is just not true, because you would still need markdown-it-py (+myst-parser) to convert to the doctree in the first place

**Reply by akhmerov** [reply]
Sorry, I fixed the description.

**Reply by chrisjsewell** [reply]
Yeh no worries, I just wanted to be very clear that `sphinx-tomyst` (formerly `sphinxcontrib-rst2myst`) is not in any way a solution to this. It acts on a defined subset of doctree nodes, and is not just generally applicable to all possible sphinx nodes (e.g. those created by extensions). It is meant as a helper tool for people who want to convert their rST file to MyST, but should definitely not be used as a core part of the build process.

(If you want to understand more, I suggest having a look at the source code
https://github.com/executablebooks/sphinx-tomyst/blob/71297ac5d0e6ff702de728b302981a4d87602754/sphinx_tomyst/writers/translator.py#L24)

**Reply by akhmerov** [reply]
Thanks, that clarified the limitations of sphinx-tomyst.

**Comment by akhmerov** [comment]
Another remark that I think is useful here is the meaning of 1-1 mapping myst ↔ ipynb. This, to the best of my understanding is a shorthand to the requirement that the following two sequences of actions produce the same result:

1. convert all myst to ipynb with jupytext
2. run all notebooks
3. build jb with execution turned off
4. inspect the resulting book as well as the notebooks

and

1. build jb from myst with execution via jupyter-cache
2. inspect the resulting book and the generated notebooks

This is more stringent than only requiring bidirectional conversion myst ↔ ipynb with jupytext. If that was the only constraint, then I believe it would be possible to support nested code cells with a lower complexity than dragging the complete stack into jupytext.

**Reply by chrisjsewell** [reply]
Yeh I think that sounds about right 👍

**Reply by akhmerov** [reply]
To be honest I am unsure in which use case the above equivalence is important. I can easily imagine developer teams utilizing:

- notebook only, no cache
- notebook only, cache always
- myst + sometimes jupytext for messing around with code, but always saving as myst

but all these scenarios lack of the workflow equivalence doesn't seem to hurt or cause unpredictable results

**Reply by chrisjsewell** [reply]
So you want to have a separate tool for each different workflow?

**Reply by akhmerov** [reply]
I am considering the following approach in mind:

1. myst-parser generates the doctree.
2. The doctree is transformed to map better on a notebook format: code cells are extracted to the top level of the doctree, references are inserted at their original locations. [note that the format here is largely the same as approach (2), except it's a sphinx doctree and not a myst file]
3. Results are plugged in from cache if enabled.
4. References to code cells are populated with outputs, glue is glued, etc.
5. sphinx-tomyst generates the notebooks to be served to the readers.

The notebooks generated this way will have the following properties:

1. all glue is glued in :tada: 
1. the notebooks are guaranteed to produce all the same results when run :heavy_check_mark:
3. they miss all custom directives that myst-parser can't handle :crying_cat_face:

---

So to answer your question, the above would apply regardless of whether the inputs are notebooks or myst files. Reading of the sources happens mostly like now.

**Reply by choldgraf** [reply]
Just a note that I would like to also take into consideration implementation effort and maintenance effort. I appreciate that there may be a globally-optimum solution out there that involves re-working everything, but we have pretty limited resources and we need to be operating with ~ the technical stack / approach that we already have (others free to disagree w/ me if you think this isn't true)

(I say this because the approach you mention above sounds like a strong departure from what we're currently doing)

**Reply by akhmerov** [reply]
You are absolutely right, this approach would require a significant restructuring of myst-nb internals.

**Reply by chrisjsewell** [reply]
> sphinx-tomyst generates the notebooks to be served to the readers.

I just told you why `sphinx-tomyst` is not adequate for this task, can you please stop suggesting it

> they miss all custom directives that myst-parser can't handle

or do you mean `sphinx-tomyst` here and not `myst-parser`, because `myst-parser` can handle all directives

**Reply by akhmerov** [reply]
So there are two things: the notebooks edited by the authors and notebooks that the readers download.

sphinx-tomyst is incapable of generating the author versions of the notebooks, no questions there. The notebooks downloaded by the users, however, serve a different purpose. They may benefit from glue substitution and they may sustain loss of custom directives.

**Reply by chrisjsewell** [reply]
Creating downloadable notebooks is not the topic of this discussion

**Reply by akhmerov** [reply]
Sure, then ignore step 5 of the build process I described above.

**Reply by chrisjsewell** [reply]
If there is not a step 5 then there is not a notebook 🤷

**Reply by akhmerov** [reply]
But myst-nb doesn't need to generate a notebook, it needs to generate a book.

Perhaps I should have stated more clearly: I am describing how myst-nb conversion process could accommodate nested code-cells if the requirement of the 1-1 correspondence was relaxed to just "myst files are editable with jupytext".

**Reply by chrisjsewell** [reply]
How are these myst files with nested cells editable with jupytext?

**Reply by akhmerov** [reply]
The user be able to execute the cells that aren't nested, and edit the nested ones inside markdown cells. The user would obviously not be able to execute those.

**Reply by chrisjsewell** [reply]
So you’re just describing option 3 then

**Reply by akhmerov** [reply]
Indeed, you're right—I've reread the description, and broadly it's that option. The details differ here though:

- There would be no new directives: the nested directives are still `code-cell`s
- The execution is still sequential, and nested directives aren't executed after all the top level ones
- nbclient still usable for execution without any modifications

So yeah, different fine print, but indeed option 3. This implementation would address some concerns you raised in https://github.com/executablebooks/meta/discussions/159#discussioncomment-94211:

- jupyter-cache/nbclient is perfectly suitable for execution
- step 5 would generate notebooks for binder/colab that contains all code in code cells.

**Reply by chrisjsewell** [reply]
As I've already pointed out, you are literally just suggesting jupyter-sphinx here

If you actually wanted to consider mapping these to physical notebooks, among a range of new issues this would generate:

> There would be no new directives: the nested directives are still code-cells

It would be hugely misleading/confusing to have some code-cells that get converted to notebook cells and some that don't. Really, it would be pointless to consider jupytext conversion, because they would not produce the same results if you run them, since you'll be skipping a bunch of code.

Also, code-cells are **NOT** directives. They don't have options like directives, they store the full 1:1 mapping of notebook metadata in YAML format

> step 5 would generate notebooks for binder/colab that contains all code in code cells.

Thats not how Binder/Colab work, they source from the Github repository. So now you would need to also commit all these generated notebooks and work out how to point links towards them.
(this on top of the fact, as I keep saying, sphinx-tomyst is not suitable for this task)

**Reply by akhmerov** [reply]
> As I've already pointed out, you are literally just suggesting jupyter-sphinx here

I don't think you mentioned jupyter-sphinx when characterizing this proposal, but this is indeed similar to what jupyter-sphinx does.

> It would be hugely misleading/confusing to have some code-cells that get converted to notebook cells and some that don't. Really, it would be pointless to consider jupytext conversion, because they would not produce the same results if you run them, since you'll be skipping a bunch of code.

Authoring experience is an important point, and it deserves careful consideration.

First of all, observe that *to some users* the situation is confusing already now because people try to nest `code-cell` directives, and observe unexpected failures. This is actually what started this discussion: executablebooks/jupyter-book#700, executablebooks/sphinx-exercise#18 (and not me :stuck_out_tongue: ). I think the users that encounter this problem use text editing as a primary interface. Plain text as the primary interface makes sense both for beginning and expert users. When authoring in plain text, the inability to nest code cells appears to confuse the authors not familiar with the internal constraints. Because it's usually 100% OK to nest directives, I fear that implementing the option (2) will not resolve this confusion.

I think jupytext conversion stays rather valuable even if nested `code-cell` directives are kept the same: notebook interface is a convenient tool to tweak the code and view the output, even if parts of the code aren't accessible. When working on my own content I routinely combine jupytext and plain text. I don't rely on the notebook execution results because I never want to commit them to git anyway.

I think that the users who insist on authoring in the notebook and directly storing the execution results wouldn't be adversely impacted neither because those wouldn't insert nested `code-cell` directives in the first place.

When you say that it would be hugely misleading to have such behavior, what kind of user and situation do you have in mind?

> Also, code-cells are **NOT** directives.

Huh? There's literally [`app.add_directive("code-cell", CodeCell)`](https://github.com/executablebooks/MyST-NB/blob/master/myst_nb/__init__.py#L170).

> Thats not how Binder/Colab work

[`echo "jb build" > .binder/postBuild`](https://mybinder.readthedocs.io/en/latest/config_files.html#postbuild-run-code-after-installing-the-environment)

**Reply by chrisjsewell** [reply]
> Huh? There's literally app.add_directive("code-cell", CodeCell)

Yeh and now look at CodeCell

https://github.com/executablebooks/MyST-NB/blob/446cf0d269a0906eacc1b88b418c7bce40ee4cee/myst_nb/__init__.py#L400

**Reply by chrisjsewell** [reply]
>  try to nest code-cell directives, and observe unexpected failures. This is actually what started this discussion: executablebooks/jupyter-book#700, executablebooks/sphinx-exercise#18

Also both of these issues reference old versions of myst-nb, before the specific warning  and link to documentation put in place

**Reply by akhmerov** [reply]
Oh wow. They certainly look a lot like directives.

**Reply by chrisjsewell** [reply]
At the end of the day, you clearly have a very specific use case / viewpoint on what you want to achieve. This is fine, but I don't get why you don't just implement all this in jupyter-sphinx?

**Reply by akhmerov** [reply]
Once again, this discussion isn't about my use case at all, and my personal educational resources are all working perfectly fine without nested code.

Still, to address your proposal: I would be happy to make jupyter-sphinx play more nicely as a stand-in replacement of myst-nb execution model, but https://github.com/executablebooks/meta/issues/152 is limiting myst-nb interoperability.

**Comment by mmcky** [comment]
Sorry to jump in later in the discussion on this. 

A comments for `sphinx-exercise`:

Would it be a reasonable assumption that `code-cell` blocks contained in admonitions are independent to the rest of the execution environment of that page? I think this could a reasonable assumption for `sphinx-exercise:solution` but probably not for `note` or other admonition types. 

From the `sphinx-exercise` viewpoint we could compile a `solutions` notebook that is a companion document (`document.ipynb`, `document_solutions.ipynb`) and all code cells in the `solutions` are added to that companion notebook which would then:

1. build companion notebook of `_solutions`
2. run both documents
3. retrieve results for `code-cell` in `solution`  block from `_solutions` notebook. 

**Alternative Build Approach:**

Another `alternative` which is closely related to original Option 2 would be to compile during the build process:

1. a content/markup file, and
2. an execution file

during the build process. The `code` file would then contain all `code blocks`/`code-cell` as found in the document from top to bottom (nested or not). The content / markup file would then get compiled by sphinx pulling in outputs from the `code` file (after execution). A hash of the `code-cell` could be used as a linking id to merge the `outputs` with `inputs`? So the key assumption would be all `code-cell` blocks are run in a linear top to bottom fashion as they appear in the `md` document (i.e. like a notebook). 

The downside as I see it from this approach is it would mix `sphinx.ast` with `execution` (if we used the `ast` to extract code-cells to produce a code-only notebook). The current approach is simpler with 1. conversion, 2. execution, 3. compile via sphinx.
